### PR TITLE
[Tailcall] Add 11 tests that are all runnable, never crash, are self checking.

### PR DESCRIPTION
**Pull request by jaykrell**
Tailcall] Add 11 tests. They all run w/o exception or stack overflow, on all platforms.

They are self-checking and report if they were tail called.
All 11 pass on Windows .NET.
2 pass with Mono on Windows or Mac.
CoreCLR not tested as it is a difficult/unknown command line environment.

Mono has problems with at least:
	virtual
	non-managed pointers
	generics
	calli
	managed references even if not to current frame

Tests originally written in C# and then run through ildasm and perl.
See https://github.com/jaykrell/mono/commits/tailcalltest2.gen.

### CODE REVIEW

File: mono/tests/tailcall/Makefile [hunk 1]
```diff
@@ -0,0 +1,54 @@
+# Tests are written in C# to be amenable to tail calls.
```
Comment (marek-safar) [inline mono/tests/tailcall/Makefile]: Do you really need custom Makefile?
Most of the dependent tools are wrong check for example https://github.com/mono/mono/blob/master/mono/tests/Makefile.am#L84 and it'll also make it harder to include in existing test suites

File: mono/tests/tailcall/Makefile [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+	$(MONO) tailcall1-tail.exe
+
+clean:
+	$(RM_F) tailcall1-tail.exe tailcall1-tail.il tailcall1.il tailcall1.exe
+
+tailcall1-tail.exe: tailcall1-tail.il
+	$(ILASM) $? /out:$@
+
+tailcall1-tail.il: tailcall1.il tailcall1.pl
+	perl tailcall1.pl < tailcall1.il > $@
```
Comment (marek-safar) [inline mono/tests/tailcall/Makefile]: really? the tests should be as simple as possible and you are patching unstable disassembled compiler output with perl script?

**Comment by jaykrell** [comment]
It is looking like splitting it up will me debug also, so I can compare e.g.  callvirt with tail.callvirt, w/o the other tests.
Another approach I was thinking was perl or C# to write out IL from scratch but using csc to produce the IL made it much easier for me, not knowing IL much yet.

**Comment by jaykrell** [comment]
But can also just comment out the others to debug..

**Comment by jaykrell** [comment]
Much eaiser to write C# and edit IL than to write IL.

Separate directory much eaiser to also make work on Windows though if I split test generation into a separate makefile, that might help.

**Comment by jaykrell** [comment]
I'll split things up but C# plus post-processing, either online or offline, is still the easiest way to write tests. Offline processing and checking .il should satisfy everyone. Replacing the ilasm/perl with something more directly IL based would be reasonable.

**Comment by jaykrell** [comment]
@monojenkins build API diff

**Comment by luhenry** [comment]
@jaykrell IL is what the runtime executes, so it's needed for all of us to understand it well.

**Pull request merged**